src/mpi/comm/comm_impl.c

```diff
@@ -623,6 +623,64 @@ int MPIR_Comm_create_group(MPIR_Comm * comm_ptr, MPIR_Group * group_ptr, int tag
     goto fn_exit;
 }
 
+int MPIR_Comm_create_from_group_impl(MPIR_Group * group_ptr, const char *stringtag,
+                                     MPIR_Info * info_ptr, MPIR_Errhandler * errhan_ptr,
+                                     MPIR_Comm ** p_newcom_ptr)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    /* This implementation assumes an internal MPI_COMM_WORLD already exist.
+     *
+     * The next implementation will relax but assume the first call to this function will
+     * either from a "world" pset or "self" pset, during former a comm world will be created.
+     *
+     * The final implementation will further relax and allow partial initializations.
+     */
+
+    /* use macro from uthash.h to calculate hash value */
+    unsigned hash;
+    int n = strlen(stringtag);
+    HASH_VALUE(stringtag, n, hash);
+
+    int tag = hash % (MPIR_Process.attrs.tag_ub);
+
+    if (MPIR_Process.comm_world) {
+        if (!MPIR_Process.comm_world->local_group) {
+            /* need create local_group for MPIR_Comm_create_group to work */
+            MPIR_Group *world_group_ptr;
+            mpi_errno = MPIR_Comm_group_impl(MPIR_Process.comm_world, &world_group_ptr);
+            MPIR_ERR_CHECK(mpi_errno);
+            MPIR_Group_release(world_group_ptr);
+        }
```

Comment:
`MPI_Comm_create_group` is essentially creating a "sub" comm. `MPI_Comm_create_from_group`, on the other hand, is potentially creating a first communicator or a communicator that gathers arbitrary processes that only partially overlaps with existing comm. Remember, in a session model, it is not given that we have a all-to-all connection or even address exchange in place. So `MPI_Comm_create_from_group` potentially is the "global" initialization.

Here of course I took the easy way out by assuming the world comm is already established during `MPI_Session_init`, so it can be simply a `MPI_Comm_create_group` from the world comm. But the code is structured partially to prepare for the next PR where we may need do the `global` initialization here. 

If we *only* support `MPI://SELF` and `MPI://WORLD`, then we do (almost) nothing here with `MPI://SELF`, and we do *global* initialization if it is `MPI://WORLD`, or we do `MPI_Comm_create_group` here if world comm is already established (current implementation). 

In advanced implementation, where we prepare to support arbitrary proc set, then we'll have to do ad-hoc semi-dynamic address exchange and communicator creation -- pushed later down the road.